File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        
        Deque<Integer> stack = new ArrayDeque<>();
        
        for(int i=0;i<s.length();i++){
            
            if(s.charAt(i)=='(') stack.push(i);
            
            else{
                
                if(stack.isEmpty()) stack.push(i);
                else{
                    if(s.charAt(stack.peek())=='(') stack.pop();
                    else stack.push(i);
                }
            }
        }
        
        if(stack.isEmpty()) return s.length();
        
        int maxLen = 0;
        int last = s.length();
        
        while(!stack.isEmpty()){
            
            int popped  = stack.pop();
            maxLen = Math.max(maxLen,last-popped-1);
            last = popped;
        }
        
        maxLen = Math.max(maxLen,last);
        
        return maxLen;
        
        
        
        
    }
}

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.py
class Solution(object):
    def longestValidParentheses(self, s):
        stack = []
        
        i = 0
        for ch in s :
            if ch=='(' : 
                stack.append(i)
            else :
                if not stack :
                    stack.append(i)
                else :
                    if s[stack[-1]]=='(' :
                        stack.pop()
                    else :
                        stack.append(i)
                        
            i+=1
        
        if(stack==None) :
            return len(s)
        maxLen = 0
        last = len(s)
        
        while stack :
            popped = stack.pop()
            maxLen = max(maxLen,last-popped-1)
            last = popped
            
        
        maxLen = max(maxLen,last)
        return maxLen
        